File: Baekjoon/Class2/39.1874 스택 수열.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const n = Number(input[0]);
const numStack = [0]; // top 접근을 위한 더미 데이터
const out = [];

let numToPush = 1;
for(let i = 1; i <= n; i++) {
    let numToPop = Number(input[i]);
    while(numStack.at(-1) < numToPop) { // pop할 숫자가 없으면
        numStack.push(numToPush++); // push하기
        out.push("+");
    }
    if (numStack.at(-1) > numToPop) { // top이 pop할 숫자 초과(실패)
        out.length = 0;
        out.push("NO");
        break;
    }
    numStack.pop(); // pop할 숫자 찾음
    out.push("-");
}
console.log(out.join("\n"));

File: Baekjoon/Class2/39.1874 스택 수열.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    stack<int> num_stack;
    string out = "";
    int n;
    cin >> n;

    num_stack.push(0); // top 접근을 위한 더미 데이터
    int num_to_push = 1;
    for (int _ = 0; _ < n; _++) { // n번 pop
        int num_to_pop;
        cin >> num_to_pop;
        while (num_stack.top() < num_to_pop) { // pop할 숫자가 없으면
            num_stack.push(num_to_push++); // push하기
            out.append("+\n");
        }
        if (num_stack.top() > num_to_pop) { // top이 pop할 숫자 초과(실패)
            out = "NO";
            break;
        }
        num_stack.pop(); // pop할 숫자 찾음
        out.append("-\n");
    }
    cout << out;
    return 0;
}